### DISCUSSION TITLE
Including __type in DataContractSerializer JSON responses

### DISCUSSION THREAD

**Discussion by qcz** [CoreWCF/CoreWCF#1486] [Q&A] [answered]
I'm trying to migrate our WCF service to CoreWCF. Almost everything works perfectly. One problem I face that I cannot find a way to configure DataContractJsonSerializer to emit type information for lists which can contain different subclasses of a given base type.

Property in DTO:
```cs
List<MyBaseClass> Properties { get; set; } // List contains MySubType, MySubType2 etc. instances
```

Old WCF service sends the following:
```json
"Properties": [
  {
    "__type": "MySubType:http://MyNamespace",
    "Name": "Name",
    "Value": "Value"
  }
],
```

while CoreWCF does not return it:
```json
"Properties": [
  {
    "Name": "Name",
    "Value": "Value"
  }
],
```

So the question basically is the following: how can I set EmitTypeInformation to AsNeeded for DataContractJsonSerializer? (I think that is the default value for WCF on .NET Framework).

Interestingly the XML output is OK, type information is added for both the old and new versions:
```xml
<Properties>
   <MyBaseClass i:type="MySubType">
     <Name>Name</Name>
     <Value>Value</Value>
   </MyBaseClass>
</Properties>
```

**Comment by qcz** [comment]
@mconnew This problem is also a deviation from the default behavior in WCF. Any insight whether the original serialization behavior is reachable somehow (even by defining a custom serializer/deserializer) in CoreWCF?

**Comment by mconnew** [comment]
I found the source of the bug. DataContractJsonSerializer used to have a public constructor which took a simple boolean to specify whether to emit type information. A true value set it to Always, a false value set it to AsNeeded. That constructor isn't available, so a different one was used which takes a DataContractJsonSerializerSettings instance instead. In the process of changing to this, a mistake was made in converting to the new semantics. It should be an easy fix, it's the tests which will take a bit of time. I'll see what I can do quickly for you.

**Comment by mconnew** [comment, marked as answer]
@qcz, could you take a look at PR #1497 to make sure the test change reflects your scenario sufficiently and also do a quick review of the changes to see if anything obvious looks wrong?

**Reply by qcz** [reply]
The solution works fine, same output as the .NET Framework version. Thanks!

However I've found an another bug while testing this: #1500.
